File: evaluate_detection_ap.py
#!/usr/bin/env python3
"""Evaluate detector average precision against P-DESTRE ground truth."""

import argparse
import glob
import os

import numpy as np
from scipy.io import loadmat

from pdestre_experiments.annotations import parse_result_basename, read_annotation_boxes
from pdestre_experiments.detection_metrics import evaluate_detection_precision, xyxy_to_xywh
from pdestre_experiments.paths import annotation_dir, detection_output_dir


def parse_args():
    parser = argparse.ArgumentParser(
        description='Compute detection AP from saved bbox .mat files.'
    )
    parser.add_argument(
        '--detection-dir',
        default=detection_output_dir(),
        help='Directory containing *_bbox.mat detector outputs',
    )
    parser.add_argument(
        '--annotation-dir',
        default=annotation_dir(),
        help='Directory with P-DESTRE annotation .txt files',
    )
    parser.add_argument(
        '--score-threshold',
        type=float,
        default=0.3,
        help='Minimum detection score kept for evaluation',
    )
    parser.add_argument(
        '--detection-scale',
        type=float,
        default=1.0,
        help='Scale factor applied to detector boxes',
    )
    parser.add_argument(
        '--gt-scale',
        type=float,
        default=4.0,
        help='Scale factor applied to ground-truth boxes',
    )
    parser.add_argument(
        '--iou-threshold',
        type=float,
        default=0.5,
        help='IoU threshold used for true-positive matching',
    )
    return parser.parse_args()


def load_detection_results(
    detection_dir,
    annotation_dir_path,
    score_threshold,
    detection_scale,
    gt_scale,
):
    detections_xywh = []
    detection_scores = []
    ground_truth_xywh = []

    bbox_files = sorted(glob.glob(os.path.join(detection_dir, '*_bbox.mat')))
    for bbox_file in bbox_files:
        video_name, frame_idx = parse_result_basename(bbox_file)
        bboxes = loadmat(bbox_file)['bboxes']
        if bboxes.ndim == 1 or bboxes.shape[0] == 0:
            detections_xywh.append(np.empty((0, 4), dtype=np.float32))
            detection_scores.append(np.empty((0,), dtype=np.float32))
        else:
            kept = bboxes[bboxes[:, 4] > score_threshold]
            xywh = xyxy_to_xywh(kept[:, :4]) / detection_scale
            detections_xywh.append(xywh.astype(np.float32))
            detection_scores.append(kept[:, 4].astype(np.float32))

        annotation_path = os.path.join(annotation_dir_path, f'{video_name}.txt')
        gt_boxes = read_annotation_boxes(annotation_path, frame_idx, scale=gt_scale)
        ground_truth_xywh.append(np.asarray(gt_boxes, dtype=np.float32))

    return detections_xywh, detection_scores, ground_truth_xywh


def main():
    args = parse_args()
    detections_xywh, detection_scores, ground_truth_xywh = load_detection_results(
        args.detection_dir,
        args.annotation_dir,
        args.score_threshold,
        args.detection_scale,
        args.gt_scale,
    )
    ap, recall, precision = evaluate_detection_precision(
        detections_xywh,
        detection_scores,
        ground_truth_xywh,
        iou_threshold=args.iou_threshold,
    )
    mean_recall = float(np.mean(recall)) if recall.size else 0.0
    mean_precision = float(np.mean(precision)) if precision.size else 0.0
    print(
        f'Average precision = {ap * 100:.2f}; '
        f'Recall = {mean_recall * 100:.2f}; '
        f'Precision = {mean_precision * 100:.2f}.'
    )


if __name__ == '__main__':
    main()

File: evaluate_detection_ap.m
% Evaluate detector average precision against P-DESTRE ground truth.
%
% Environment variables:
%   PDESTRE_DETECTION_OUTPUT_DIR
%   PDESTRE_ANNOTATION_DIR

clear;

vis_path = getenv('PDESTRE_DETECTION_OUTPUT_DIR');
if isempty(vis_path)
    vis_path = 'outputs/visible_PDestre_bottom_exp_out';
end

GT_path = getenv('PDESTRE_ANNOTATION_DIR');
if isempty(GT_path)
    GT_path = 'data/P_Destre/annotation';
end

D = dir(fullfile(vis_path, '*_bbox.mat'));
N = size(D, 1);
resize_factor = 1;
resize_factorGT = 4;

annotations = [];
scores = [];
fprintf('Starting...');
for i = 1 : N
    bb_file = fullfile(D(i).folder, D(i).name);
    bboxes = load(bb_file);
    bboxes = bboxes.bboxes;
    if size(bboxes, 1) > 0
        a = bboxes(bboxes(:, 5) > 0.3, 1:4);
        a(:, 3:4) = a(:, 3:4) - a(:, 1:2);
        annotations{i} = a / resize_factor;
        scores{i} = bboxes(bboxes(:, 5) > 0.3, 5);
    else
        annotations{i} = [];
        scores{i} = [];
    end

    annotation_array = [];
    [~, name, ~] = fileparts(D(i).name);
    parts = split(name, '_');
    frame_idx = parts{3};
    fileN = fullfile(GT_path, [parts{1}, '.txt']);
    fileID = fopen(fileN, 'r');
    line = fgetl(fileID);
    if line == -1
        fclose(fileID);
        continue;
    end
    while ischar(line)
        info = split(line, ',');
        frame = info{1};
        if strcmp(frame, frame_idx)
            coor_double = [str2double(info{3}), str2double(info{4}), str2double(info{5}), str2double(info{6})];
            coor_double(coor_double < 0) = 1;
            coor_double = coor_double / resize_factorGT;
            annotation_array = [annotation_array; coor_double]; %#ok<AGROW>
        end
        line = fgetl(fileID);
    end
    annotationsGT{i} = annotation_array;
    fclose(fileID);
end

visible_table = cell2table([annotations', scores']);
GT_table = cell2table(annotationsGT');

[ap, recall, precision] = evaluateDetectionPrecision(visible_table, GT_table);
fprintf('Average precision = %.2f; Recall = %.2f; Precision = %.2f. \n', ap * 100, mean(recall) * 100, mean(precision) * 100);
